index.js

```diff
@@ -1,135 +1,122 @@
-
-  mkdirp.sync(path.join('projects', pid))
-  var dbdir = path.join('projects', pid, 'db')
-  var osm = Osm({ dir: dbdir, encryptionKey: pid })
-  var media = Blob(path.join('projects', pid, 'media'))
-
-  projectCores[pid] = new Mapeo(osm, media, { internetDiscovery: true })
-  projectCores[pid].sync.setName('mapeo-web') // TODO: some way for the operator to provide this
-  projectCores[pid].sync.listen()
-  projectCores[pid].sync.join(pid)
+module.exports = {
+  create
+}
 
-  // provides a reverse map of HASH => projectId
-  osm.core.use('filtermap', filtermap(pid))
+function create (opts) {
+  const mapeoWeb = new MapeoWeb(opts)
 
-  return projectCores[pid]
-}
+  mapeoWeb.initRoutes()
 
-function getProject (pid) {
-  return projectCores[pid]
+  return mapeoWeb
 }
 
-function removeProject (pid, cb) {
-  var core = getProject(pid)
-  if (!core) return process.nextTick(cb)
-  core.close(function () {
-    fs.rename(path.join('projects', pid), path.join('projects', 'dead-' + String(Math.random()).slice(2)), cb)
-    delete projectCores[pid]
-  })
-}
+class MapeoWeb {
+  constructor ({ logger = pino(PINO_OPTS), ...opts } = {}) {
+    const permissions = new Permissions({ ...opts, logger })
+    const multiMapeo = new MultiMapeo({ ...opts, logger })
+    const fastify = makeFastify({ logger })
+
+    this.logger = logger
+    this.permissions = permissions
+    this.multiMapeo = multiMapeo
+    this.fastify = fastify
+  }
+
+  initRoutes () {
+    this.fastify.register(require('fastify-websocket'))
+
+    // Leaving room for other major minor and patch versions.
+    this.fastify.get('/replicate/v1/:discoveryKey', { websocket: true }, (connection, req, params) => {
+      const { discoveryKey } = params
+
+      this.permissions.getProjectKeyForDiscoveryKey(discoveryKey, (err, projectKey) => {
+        if (err || !projectKey) {
+          return connection.end('Invalid Key')
+        }
+        this.multiMapeo.replicate(connection, req, projectKey)
+      })
+    })
 
-// Q: how to avoid needing to scan every core?
-function getProjectIdAndFilterIdFromExportId (exportId, cb) {
-  var tasks = Object.values(projectCores)
-    .map(function (core) {
-      return function (cb) {
-        core.osm.ready(function () {
-          core.osm.core.api.filtermap.get(exportId, function (err, values) {
-            if (err && err.notFound) err = null
-            if (err) return cb(err)
-            if (!values || !values.length) return cb()
-            var filter = values[0].value
-            cb(null, [core, filter])
-          })
+    this.fastify.post('/projects/', {
+      schema: {
+        body: {
+          type: 'object',
+          properties: {
+            projectKey: {
+              type: 'string',
+              minLength: 64,
+              maxLength: 64
+            }
+          }
+        }
+      },
+      handler: (req, reply) => {
+        const { projectKey } = req.body
+        this.permissions.addProjectKey(projectKey, (err) => {
```

Comment:
Need to check that the project key is valid hex (`/[^0-9a-f]/.test(projectKey)`) because otherwise the server will crash with an uncaught error (because we do not catch the error that can be thrown in the discoveryKey function).

Reply:
The error already throws properly, so I'm not sure if anything needs to be added here.